### PR TITLE
kvclient: fix error handling on proxy requests

### LINKED ISSUE #121168: sql/logictest: TestParallel failed
sql/logictest.TestParallel [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14567878?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14567878?buildTab=artifacts#/) on release-23.2.3-rc @ [d57701520e29e5226a1e3832b02ecf2810da40bd](https://github.com/cockroachdb/cockroach/commits/d57701520e29e5226a1e3832b02ecf2810da40bd):

```
	NOTICE: .ClusterSettings() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().ClusterSettings() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/sql/logictest/parallel_test.go:178: ((*parallelTest).setup)
	NOTICE: .ClusterSettings() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().ClusterSettings() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/setup with config : 2 tests, 0 failures
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 1 tests, 0 failures
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 0 tests, 0 failures
    parallel_test.go:85: 
        testdata/parallel_test/subquery_retry_multinode/txn:9: 
        expected success, but found
        (40003) result is ambiguous: error=TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=7f6e870c key=/Tenant/10/Table/106/1/954965585192878084/0 iso=Serializable pri=0.02014255 epo=0 ts=1711502768.527534781,1 min=1711502768.388751022,0 seq=3} lock=true stat=PENDING rts=1711502768.516643055,0 wto=false gul=1711502768.888751022,0 [propagate] (last error: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=7f6e870c key=/Tenant/10/Table/106/1/954965585192878084/0 iso=Serializable pri=0.02014255 epo=0 ts=1711502768.527534781,2 min=1711502768.388751022,0 seq=3} lock=true stat=PENDING rts=1711502768.516643055,0 wto=false gul=1711502768.888751022,0)
        dist_sender.go:3050: in sendToReplicas()
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 0 tests, 0 failures
    parallel_test.go:85: 
        testdata/parallel_test/subquery_retry_multinode/txn:9: 
        expected success, but found
        (40003) result is ambiguous: error=TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=9f294017 key=/Tenant/10/Table/106/1/954965585213620227/0 iso=Serializable pri=0.00827606 epo=0 ts=1711502768.527534781,2 min=1711502768.389020036,0 seq=3} lock=true stat=PENDING rts=1711502768.527534781,0 wto=false gul=1711502768.889020036,0 [propagate] (last error: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=9f294017 key=/Tenant/10/Table/106/1/954965585213620227/0 iso=Serializable pri=0.00827606 epo=0 ts=1711502768.527534781,2 min=1711502768.389020036,0 seq=3} lock=true stat=PENDING rts=1711502768.527534781,0 wto=false gul=1711502768.889020036,0)
        dist_sender.go:3050: in sendToReplicas()
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 1 tests, 0 failures
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 0 tests, 0 failures
    parallel_test.go:85: 
        testdata/parallel_test/subquery_retry_multinode/txn:9: 
        expected success, but found
        (40003) result is ambiguous: error=TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=27785b27 key=/Tenant/10/Table/106/1/954965585221419013/0 iso=Serializable pri=0.02867952 epo=0 ts=1711502768.527534781,2 min=1711502768.388961654,0 seq=3} lock=true stat=PENDING rts=1711502768.527132080,0 wto=false gul=1711502768.888961654,0 [propagate] (last error: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=27785b27 key=/Tenant/10/Table/106/1/954965585221419013/0 iso=Serializable pri=0.02867952 epo=0 ts=1711502768.527534781,2 min=1711502768.388961654,0 seq=3} lock=true stat=PENDING rts=1711502768.527132080,0 wto=false gul=1711502768.888961654,0)
        dist_sender.go:3050: in sendToReplicas()
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/final with config : 2 tests, 0 failures
    parallel_test.go:253: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/556c3b72a9585bf329eaad0cafbabdf0/logTestParallel_subquery_retry_multinode1358601955
I240327 01:26:08.850159 12983 sql/logictest/parallel_test.go:258  [-] 2  1 out of 3 parallel tests failed
    --- FAIL: TestParallel/subquery_retry_multinode (33.63s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #120974 sql/logictest: TestParallel failed [C-test-failure O-robot T-sql-queries branch-master release-blocker]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParallel.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

Jira issue: CRDB-37105

### PR DESCRIPTION
**Pull request by andrewbaptist**
Previously on a proxy request, the error was extracted and wrapped into a ProxyFailedError incorrectly. It should only be moved into a ProxyFailedError if it is a non-remote error. This PR cleans up the handling of any proxy requests (requests with ProxyRangeInfo) set on them.

Epic: none
Fixes: #121168

Release note: None

Release Justification: This change prevents errors on proxy requests from incorrectly being converted to ambiguous errors on the sender.

### CODE REVIEW

**Review by arulajmani — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121317)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAQ5QREBQq6VjcymFl):*
nit: this was subtle to debug, so I think the commit message could use more details. Also, consider adding some words about the consequence of this wrapping as well.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2099 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAPOm82p2VugkMfsNh:-NuAPOm82p2VugkMfsNi:b-xfwzfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab61acf332cf1c13053a68ee7b3d061e3be5587b/pkg/kv/kvclient/kvcoord/dist_sender.go#L2099)):*
> ```Go
> 		// If ProxyRangeInfo is set, this is a proxy request from a remote
> 		// client. Proxy requests are retried on the remote client and we don't
> 		// need to look at the reply.Error. The client that initiated this proxy
> ```

nit: Can we improve this entire comment a bit? I mostly follow what you're trying to say here because we discussed this over a call yesterday, but I think it might be hard to come back to this in the future or for a reader not intimately familiar with the details. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuANeTT7xm45SUUs4qk:-NuANeTU520vW45qh2SC:b7y8d2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab61acf332cf1c13053a68ee7b3d061e3be5587b/pkg/kv/kvclient/kvcoord/dist_sender.go#L2109)):*
> ```Go
> 				// Wrap the error in a kvpb.Error to match the interface of
> 				// Send. It is immediately unwrapped in node.maybeProxyRequest.
> 				return response{pErr: kvpb.NewError(err)}
> ```

Does this mean we could be sending `sendErrors` back to the remote client? Is that something we currently/want to handle?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAPmVVAGV2Y150NW8-:-NuAPmVVAGV2Y150NW80:b-o26sdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab61acf332cf1c13053a68ee7b3d061e3be5587b/pkg/kv/kvclient/kvcoord/dist_sender.go#L2111)):*
> ```Go
> 				return response{pErr: kvpb.NewError(err)}
> 			} else {
> 				log.VEventf(ctx, 2, "proxy request succeeded with %s", reply)
> ```

The proxy request may have failed if `reply.Error != nil`, right?

**Review by andrewbaptist — COMMENTED** [review]
TFTR - I've delved deeper into the woods with this, and this fix is correct and more accurate, but still not 100% sufficient. We should get this in as it fixes almost all errors, but there is still a small class of problems due to underspecified error behavior in sendToReplicas.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-:-NuAbO0q31XgsWp_zuS-:b-i0khwr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAQ5QREBQq6VjcymFl):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: this was subtle to debug, so I think the commit message could use more details. Also, consider adding some words about the consequence of this wrapping as well.
</blockquote></details>

I added and cleaned up the commit message
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2099 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAPOm82p2VugkMfsNh:-NuA_SlmBBcvqOx1T8wx:b-blranm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab61acf332cf1c13053a68ee7b3d061e3be5587b/pkg/kv/kvclient/kvcoord/dist_sender.go#L2099)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Can we improve this entire comment a bit? I mostly follow what you're trying to say here because we discussed this over a call yesterday, but I think it might be hard to come back to this in the future or for a reader not intimately familiar with the details.
</blockquote></details>

I mostly rewrote the comment. This is a complex topic, and I am adding a large comment to the top of `sendToReplicas` as well to help clarify this in a future commit. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuANeTT7xm45SUUs4qk:-NuA_qpS3Zopzp_gGStI:bzeyiev) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab61acf332cf1c13053a68ee7b3d061e3be5587b/pkg/kv/kvclient/kvcoord/dist_sender.go#L2109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Does this mean we could be sending `sendErrors` back to the remote client? Is that something we currently/want to handle?
</blockquote></details>

Yes - we have to do something with these errors. The options are
1) retry here - but that means we are using different information to make our call than the client did
2) drop the error - this is bad because this request didn't succeed
3) send the error back to the client and let them handle it.  This is what we want to do. 

I added some clarifying text to the comment above as well.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAPmVVAGV2Y150NW8-:-NuA_kRo7UgXEQUDUu7H:blynkbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab61acf332cf1c13053a68ee7b3d061e3be5587b/pkg/kv/kvclient/kvcoord/dist_sender.go#L2111)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The proxy request may have failed if `reply.Error != nil`, right?
</blockquote></details>

Correct - I changed this to 'completed' instead of succeeded.

**Review by arulajmani — COMMENTED** [review]
Mostly LGTM modulo nits.

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-:-NuWVIYcCGAZOxPHr55v:b-631y7p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuAQ5QREBQq6VjcymFl):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

I added and cleaned up the commit message
</blockquote></details>

Thanks!
___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWSCWF2gUvjbQs4o-C):*
nit: s/SendError/sendError/g (here, and everywhere else)
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWSqsf36Rh5OaUrsIX:-NuWSqsf36Rh5OaUrsIY:b-h3i0na) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2119)):*
> ```Go
> 				reply = &kvpb.BatchResponse{}
> 				if IsSendError(err) {
> 					// SendErrors don't escape from DistSender, we treat this as
> ```

nit: consider rewording this to something similar to:

```
// sendErrors shouldn't[1] escape from the DistSender. If routing the request resulted in a sendError,
// we intercept it here and replace it with a ProxyFailedWithSendError.
//
// [1] sendErrors are a mechanism for sendToReplicas to communicate back that the routing information
// used to route the request was stale. Therefore, the cache needs to be flushed and the request should be
// retried with fresher information. However, this is specific to the proxy node's range cache -- not the remote
// client's range cache. As such, returning the sendError back to the remote client is non-sensical.
```
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2757 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWUD2B9QptxjvIMGNt:-NuWUD2B9QptxjvIMGNu:bd4gv6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2757)):*
> ```Go
> 				// The server proxy attempt resulted in an error, likely a
> 				// communication error. Convert the error on the batch request
> 				// to a Send error and handle it as if we hit this error
> ```

Is this comment stale?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3045 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWURErDl_aT1Se5ply:-NuWUREs4PADz7kjct0l:b-zc63it) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L3045)):*
> ```Go
> 								ctx, 2, "transport incompatible with updated routing; bailing early",
> 							)
> 							return nil, noMoreReplicasErr(
> ```

Did we need this change?
___
*[`pkg/server/node.go` line 1542 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWUqjZDkxkYtDzirEo:-NuWUqjZDkxkYtDzirEp:b-bp7yy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/server/node.go#L1542)):*
> ```Go
> 		return nil
> 	}
> 	// On a proxy request, it is rare to get here, wrapping normally happens in
> ```

nit: Consider switching to "It is rare to get here on a proxy request because wrapping..."

**Review by andrewbaptist — COMMENTED** [review]
TFTR! I pulled the one change out and cleaned up the comments. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-:-NuZ2iQk9dWvZJarF_Tj:b-xip3v3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWSCWF2gUvjbQs4o-C):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/SendError/sendError/g (here, and everywhere else)
</blockquote></details>

I changed everywhere. Thanks!
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWSqsf36Rh5OaUrsIX:-NuZ12aT39DxRl7ULDIf:bofwrhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2119)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider rewording this to something similar to:

```
// sendErrors shouldn't[1] escape from the DistSender. If routing the request resulted in a sendError,
// we intercept it here and replace it with a ProxyFailedWithSendError.
//
// [1] sendErrors are a mechanism for sendToReplicas to communicate back that the routing information
// used to route the request was stale. Therefore, the cache needs to be flushed and the request should be
// retried with fresher information. However, this is specific to the proxy node's range cache -- not the remote
// client's range cache. As such, returning the sendError back to the remote client is non-sensical.
```
</blockquote></details>

Thanks! I used your wording!
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2757 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWUD2B9QptxjvIMGNt:-NuZ2EmW0r4oHE8M4AGF:b-r5jjn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2757)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this comment stale?
</blockquote></details>

It was a little stale, I clarified it.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3045 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWURErDl_aT1Se5ply:-NuZ2IxD9nYpq1T-nTiU:b-frn0o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L3045)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did we need this change?
</blockquote></details>

We need this change, but not in this PR. I'm creating another PR to address this issue. 
___
*[`pkg/server/node.go` line 1542 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWUqjZDkxkYtDzirEo:-NuZ2gsa4xBkhbQ0Zg14:b499q7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/server/node.go#L1542)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Consider switching to "It is rare to get here on a proxy request because wrapping..."
</blockquote></details>

Changed

**Review by arulajmani — APPROVED** [review]
:lgtm_strong: 

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-:-NuZCnedBhtwzQiZXr3q:b-uwkowd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWSqsf36Rh5OaUrsIX:-NuZC42r4QEgRMHR3yu4:b-e86ef2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2119)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Thanks! I used your wording!
</blockquote></details>

Thanks for updating the comment!
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3045 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121317#-NuWURErDl_aT1Se5ply:-NuZCICL87fVDDkzv0Gk:b-1jtlqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db114cf37200000aa7fba48a2719b44dc17197a/pkg/kv/kvclient/kvcoord/dist_sender.go#L3045)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

We need this change, but not in this PR. I'm creating another PR to address this issue.
</blockquote></details>

I'm probably missing the motivation here. Thanks for reverting the change in this PR though -- we shouldn't hold this one up given the CI fallout.

**Comment by andrewbaptist** [comment]
TFTR!

bors r=arulajmani

**Pull request merged**
